polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -84,14 +88,56 @@ bool isCandidate(LoopLikeOpInterface loop) {
   return true;
 }
 
+/// Get the size of unused shared local memory arena in bytes.
+unsigned getLocalMemRemain(Operation &module) {
+  assert(module.hasTrait<OpTrait::SymbolTable>() &&
+         "Expecting module with SymbolTable trait");
+  unsigned localMemRemain = LocalMemSize;
+  module.walk([&localMemRemain](memref::GlobalOp global) {
+    MemRefType memRefTy = global.getType();
+    auto memSpace =
+        dyn_cast_or_null<sycl::AccessAddrSpaceAttr>(memRefTy.getMemorySpace());
+    if (!memSpace || memSpace.getValue() != sycl::AccessAddrSpace::LocalAccess)
+      return WalkResult::advance();
+    unsigned globalSize =
+        memRefTy.getElementTypeBitWidth() * memRefTy.getNumElements() / 8;
+    if (globalSize >= localMemRemain) {
+      localMemRemain = 0;
+      return WalkResult::interrupt();
+    }
+    localMemRemain = localMemRemain - globalSize;
+    return WalkResult::advance();
+  });
+  return localMemRemain;
+}
+
+/// Create a GlobalOp for workgroup local memory.
+memref::GlobalOp getWGLocalMem(LoopLikeOpInterface loop) {
+  Operation *module = loop->getParentWithTrait<OpTrait::SymbolTable>();
+  unsigned localMemRemain = getLocalMemRemain(*module);
+  assert(localMemRemain != 0 && "Expecting local memory");
+  std::string name = "WGLocalMem";
+  polygeist::getUniqueSymbolName(name, module);
+  OpBuilder globalBuilder(module->getRegion(0));
+  auto memRefTy = MemRefType::get(
+      localMemRemain, globalBuilder.getI8Type(), MemRefLayoutAttrInterface(),
```

Comment:
Do I understand correctly that we claim all remaining local memory here? 

This can have a negative impact on performance, in particular if we don't need all of it. 

If one work-group needs all of the local memory, that impacts occupancy of the core, as no other (or at least fewer) work-groups can be scheduled on the same core. Generally, we should only claim as much local memory as we actually need for optimization (and even then there is a heuristic tradeoff between internalization and occupancy).

Reply:
The problem we are facing is the local memory needed is not a compile time known constant, and creating a global memref require a compile time known constant.

Reply:
We need to use static local memory so we have to choose at compile time how much to potentially use. We need to claim it only if the compiler find an opportunity to promote at least one memory access to local memory. So kernels that do not have any opportunity will not suffer in performance at all. For kernels that do have an opportunity is likely that the array being promoted is larger that 32KB (and 32 KB is not actually the size of the real local memory on the GPU, is the minimum it could have AFAIK). So we are being conservative from that point of view. And a loop is likely to have more than one array being promoted to local memory.

We could allocate a portion of the minimum (like half) to be even more conservative and introduce an option to change the percentage of local mem that can be used by this optimization.

Reply:
There's a way for the user to specify this or the compiler to fix this size: 
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes

Specifically, `reqd_work_group_size` is of interest here. If the user specified this, the work-group size is known. 

If the user did not specify this, the compiler could set it (to some meaningful value for that platform) and attach the attribute. At runtime, the user can query this value and an exception would be thrown if the user-provided local range during submission deviates from this value.

Reply:
How about...
when user specify `reqd_work_group_size`, 
        => compile uses that information to obtain local memory.
otherwise
        => when `nd_item` is passed, 
                 => obtain 32KB local memory
        => when non-`nd_item` is passed, (assuming user cannot specify local range)
                 => compiler add `reqd_work_group_size`

Reply:
> How many of the benchmarks we currently care about do even use `nd_item`?

The one written by Ettore is using `nd_item`, which can be changed.

We just need to ensure we don't change the behavior of a program after an optimization, e.g., thrown as compiler adds an attribute.

Reply:
> How about...
> 
> ```
> when user specify `reqd_work_group_size`, 
>         => compiler uses that information to obtain local memory.
> otherwise
>         => when `nd_item` is passed, 
>                  => compiler obtains 32KB local memory
>         => when non-`nd_item` is passed, (assuming user cannot specify local range)
>                  => compiler adds `reqd_work_group_size` 
> ```

That sounds like a reasonable approach, and would allow us to evaluate the impact on occupancy for the `nd_item` version with benchmarks 👍

Reply:
> We just need to ensure we don't change the behavior of a program after an optimization, e.g., thrown as compiler adds an attribute.

Yes, I agree. 

Would be interesting to at some point investigate how many actual benchmarks use `nd_item`. Using `item` gives the runtime/driver the chance to choose a value. `nd_item` would mainly be needed if group synchronization or algorithms are needed, or the algorithm requires a certain group size. In particular in the latter case, would also be interesting to investigate if there's any opportunity for this optimization there.

Reply:
Created https://github.com/intel/llvm/pull/9712 to add support of reqd_work_group_size in cgeist.

Reply:
```
        => when `nd_item` is passed, 
                 => compiler obtains 32KB local memory
        => when non-`nd_item` is passed, (assuming user cannot specify local range)
                 => compiler adds `reqd_work_group_size` 
```
is done.